cpp/log/cert.cc

```diff
@@ -546,12 +588,337 @@ Cert::Status Cert::ExtensionStructure(int extension_nid,
       LOG(WARNING) << "Corrupt extension data";
       LOG_OPENSSL_ERRORS(WARNING);
     }
+
     return FALSE;
   }
 
   return TRUE;
 }
 
+bool IsRedactedHost(const string& hostname) {
+  // Split the hostname on '.' characters
+  const vector<string> tokens(util::split(hostname, '.'));
+
+  for (const string& str : tokens) {
+    if (str == "?") {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool IsValidRedactedHost(const string& hostname) {
+  // Split the hostname on '.' characters
+  const vector<string> tokens(util::split(hostname, '.'));
+
+  // Enforces the following rules: '?' must be to left of non redactions
+  // If first label is '*' then treat it as if it was a redaction
+  bool can_redact = true;
+  for (int pos = 0; pos < tokens.size(); ++pos) {
+    if (tokens[pos] == "?") {
+      if (!can_redact) {
+        return false;
+      }
+    } else {
+      // Allow a leading '*' for redaction but once we've seen anything else
+      // forbid further redactions
+      if (tokens[pos] != "*") {
+        can_redact = false;
+      } else if (pos > 0) {
+        // '*' is only valid at the left
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+// Helper method for validating V2 redaction rules. If it returns true
+// then the result in status is final.
+bool Cert::ValidateRedactionSubjectAltNameAndCN(int *dns_alt_name_count,
```

Comment:
There are a couple of large (100+ line) methods added here, can they be broken up a bit to make it easier to read?